src/main/java/net/consensys/orion/http/handler/send/SendHandler.java

```diff
@@ -66,110 +67,109 @@ public SendHandler(
       final Storage<PrivacyGroupPayload> privacyGroupStorage,
       final Storage<QueryPrivacyGroupPayload> queryPrivacyGroupStorage,
       final ConcurrentNetworkNodes networkNodes,
-      final HttpContentType contentType,
+      final SendRequestParser parser,
       final Config config) {
     this.enclave = enclave;
     this.storage = storage;
     this.privacyGroupStorage = privacyGroupStorage;
     this.queryPrivacyGroupStorage = queryPrivacyGroupStorage;
     this.nodeKeys = Arrays.asList(enclave.nodeKeys());
     this.networkNodes = networkNodes;
-    this.contentType = contentType;
+    this.parser = parser;
     this.httpClient = NodeHttpClientBuilder.build(vertx, config, 1500);
   }
 
   @Override
   public void handle(final RoutingContext routingContext) {
-    final SendRequest sendRequest;
-    if (contentType == JSON) {
-      sendRequest = Serializer.deserialize(JSON, SendRequest.class, routingContext.getBody().getBytes());
-    } else {
-      sendRequest = binaryRequest(routingContext);
-    }
-    log.debug(sendRequest);
+    final SendRequest sendRequest = parser.parse(routingContext);
+    final PublicKey fromKey = readPublicKey(sendRequest);
 
-    if ((sendRequest.to() == null || sendRequest.to().length == 0) && !sendRequest.privacyGroupId().isPresent()) {
-      sendRequest.setTo(new String[] {sendRequest.from().orElse(null)});
-    }
-    if (!sendRequest.isValid()) {
-      throw new OrionException(OrionErrorCode.INVALID_PAYLOAD);
+    if (sendRequest.privacyGroupId().isEmpty()) {
+      handleLegacySendRequest(routingContext, sendRequest, fromKey);
+    } else if (sendRequest.privacyGroupId().isPresent()) {
+      handlePrivacyGroupSendRequest(routingContext, sendRequest, fromKey);
     }
+  }
 
+  private PublicKey readPublicKey(SendRequest sendRequest) {
     log.debug("reading public keys from SendRequest object");
     // read provided public keys
-    final Box.PublicKey fromKey = sendRequest.from().map(enclave::readKey).orElseGet(() -> {
+    return sendRequest.from().map(enclave::readKey).orElseGet(() -> {
       if (nodeKeys.isEmpty()) {
         throw new OrionException(OrionErrorCode.NO_SENDER_KEY);
       }
       return nodeKeys.get(0);
     });
+  }
```

Comment:
Wondering if this is actually part of the parsing - i.e. given the parser injects the "To" (if its missing) - maybe it could do this too - and thus if no key is available, it fails parsing, rather than "handling" ... maybe?